File: poc/MultiViewFeaturePoC/MultiViewFeaturePoC/Services/WindowManagerService.cs
﻿using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;

using Windows.ApplicationModel.Core;
using Windows.UI.Core;
using Windows.UI.ViewManagement;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace MultiViewFeaturePoC.Services
{
    public delegate void ViewClosedHandler(ViewLifetimeControl viewControl, EventArgs e);

    public class WindowManagerService
    {
        // TODO WTS: See more details about showing multiple views for an app
        // https://docs.microsoft.com/windows/uwp/design/layout/show-multiple-views
        private static WindowManagerService _current;
        public static WindowManagerService Current => _current ?? (_current = new WindowManagerService());

        // Contains all the opened secondary views.
        public readonly ObservableCollection<ViewLifetimeControl> SecondaryViews = new ObservableCollection<ViewLifetimeControl>();

        public int MainViewId { get; private set; }

        public CoreDispatcher MainDispatcher { get; private set; }

        public void Initialize()
        {
            MainViewId = ApplicationView.GetForCurrentView().Id;
            MainDispatcher = Window.Current.Dispatcher;
        }

        // TODO WTS: Displays a view as a standalone
        // You can use the resulting ViewLifeTileControl to interact with the new window.
        public async Task<ViewLifetimeControl> TryShowAsStandaloneAsync(string windowTitle, Type pageType)
        {
            ViewLifetimeControl viewControl = await CreateViewLifetimeControlAsync(windowTitle, pageType);
            SecondaryViews.Add(viewControl);
            viewControl.StartViewInUse();
            var viewShown = await ApplicationViewSwitcher.TryShowAsStandaloneAsync(viewControl.Id, ViewSizePreference.Default, ApplicationView.GetForCurrentView().Id, ViewSizePreference.Default);
            viewControl.StopViewInUse();
            return viewControl;
        }


        // Displays a view as a standalone view in the desired view mode
        public async Task<ViewLifetimeControl> TryShowAsViewModeAsync(string windowTitle, Type pageType, ApplicationViewMode viewMode = ApplicationViewMode.Default)
        {
            ViewLifetimeControl viewControl = await CreateViewLifetimeControlAsync(windowTitle, pageType);
            SecondaryViews.Add(viewControl);
            viewControl.StartViewInUse();
            var viewShown = await ApplicationViewSwitcher.TryShowAsViewModeAsync(viewControl.Id, viewMode);
            viewControl.StopViewInUse();
            return viewControl;
        }


        private async Task<ViewLifetimeControl> CreateViewLifetimeControlAsync(string windowTitle, Type pageType)
        {
            ViewLifetimeControl viewControl = null;

            await CoreApplication.CreateNewView().Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                viewControl = ViewLifetimeControl.CreateForCurrentView();
                viewControl.Title = windowTitle;
                viewControl.StartViewInUse();
                var frame = new Frame();
                frame.Navigate(pageType, viewControl);
                Window.Current.Content = frame;
                Window.Current.Activate();
                ApplicationView.GetForCurrentView().Title = viewControl.Title;
            });

            return viewControl;
        }

        public bool IsWindowOpen(string windowTitle) => SecondaryViews.Any(v => v.Title == windowTitle);
    }
}

File: templates/Features/MultiViewVB/Services/WindowManagerService.vb
﻿Imports Windows.ApplicationModel.Core
Imports Windows.UI.Core

Namespace Services

    Public Delegate Sub ViewClosedHandler(ByVal viewControl As ViewLifetimeControl, ByVal e As EventArgs)

    Public Class WindowManagerService

        ' TODO WTS: See more details about showing multiple views for an app
        ' https://docs.microsoft.com/windows/uwp/design/layout/show-multiple-views
        Private Shared _current As WindowManagerService

        Public Shared ReadOnly Property Current As WindowManagerService
            Get
                Return If(_current, (__InlineAssignHelper(_current, New WindowManagerService())))
            End Get
        End Property

        ' Contains all the opened secondary views.
        Public Property SecondaryViews As ObservableCollection(Of ViewLifetimeControl) = New ObservableCollection(Of ViewLifetimeControl)()

        Public Property MainViewId As Integer

        Public Property MainDispatcher As CoreDispatcher

        Public Sub Initialize()
            MainViewId = ApplicationView.GetForCurrentView().Id
            MainDispatcher = Window.Current.Dispatcher
        End Sub

        ' TODO WTS: Displays a view as a standalone
        ' You can use the resulting ViewLifeTileControl to interact with the new window.
        Public Async Function TryShowAsStandaloneAsync(ByVal windowTitle As String, ByVal pageType As Type) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Await CreateViewLifetimeControlAsync(windowTitle, pageType)
            SecondaryViews.Add(viewControl)
            viewControl.StartViewInUse()
            Dim viewShown = Await ApplicationViewSwitcher.TryShowAsStandaloneAsync(viewControl.Id, ViewSizePreference.[Default], ApplicationView.GetForCurrentView().Id, ViewSizePreference.[Default])
            viewControl.StopViewInUse()
            Return viewControl
        End Function

        ' Displays a view as a standalone view in the desired view mode
        Public Async Function TryShowAsViewModeAsync(ByVal windowTitle As String, ByVal pageType As Type, ByVal Optional viewMode As ApplicationViewMode = ApplicationViewMode.[Default]) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Await CreateViewLifetimeControlAsync(windowTitle, pageType)
            SecondaryViews.Add(viewControl)
            viewControl.StartViewInUse()
            Dim viewShown = Await ApplicationViewSwitcher.TryShowAsViewModeAsync(viewControl.Id, viewMode)
            viewControl.StopViewInUse()
            Return viewControl
        End Function

        Private Async Function CreateViewLifetimeControlAsync(ByVal windowTitle As String, ByVal pageType As Type) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Nothing
            Await CoreApplication.CreateNewView().Dispatcher.RunAsync(CoreDispatcherPriority.Normal, Sub()
                viewControl = ViewLifetimeControl.CreateForCurrentView()
                viewControl.Title = windowTitle
                viewControl.StartViewInUse()
                Dim frame = New Frame()
                frame.Navigate(pageType, viewControl)
                Window.Current.Content = frame
                Window.Current.Activate()
                ApplicationView.GetForCurrentView().Title = viewControl.Title
            End Sub)
            Return viewControl
        End Function

        Public Function IsWindowOpen(ByVal windowTitle As String) As Boolean
            Return SecondaryViews.Any(Function(v) v.Title = windowTitle)
        End Function

        '<Obsolete("Please refactor code that uses this function, it is a simple work-around to simulate inline assignment in VB!")>
        Private Shared Function __InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
            target = value
            Return value
        End Function
    End Class
End Namespace
